src/verifier.rs

```diff
@@ -0,0 +1,406 @@
+use std::alloc::handle_alloc_error;
+
+use anyhow::Result;
+
+use crate::{
+    AffinePoint, Circuit, Curve, Field, HaloCurve, msm_execute, msm_execute_parallel,
+    msm_precompute, NUM_ROUTED_WIRES, NUM_WIRES, ProjectivePoint, Proof,
+};
+use crate::fft::{fft_precompute, ifft_with_precomputation_power_of_2};
+use crate::partition::get_subgroup_shift;
+use crate::plonk_challenger::Challenger;
+use crate::plonk_gates::evaluate_all_constraints;
+use crate::plonk_proof::OpeningSet;
+use crate::plonk_util::{eval_poly, halo_g, halo_n, pedersen_hash, powers, reduce_with_powers};
+
+const SECURITY_BITS: usize = 128;
+
+pub struct VerificationKey<C: Curve> {
+    selector_commitments: Vec<AffinePoint<C>>,
+    sigma_commitments: Vec<AffinePoint<C>>,
+    degree_log: usize,
+    degree_pow: usize,
+}
+
+pub fn verify_proof_circuit<C: HaloCurve, InnerC: HaloCurve<BaseField=C::ScalarField>>(
+    public_inputs: &[C::ScalarField],
```

Comment:
Could we extract public inputs from the proof, something like this?

    (0..circuit.num_public_inputs).map(|i| proof.o_public_inputs[i / NUM_WIRES].o_wires[i % NUM_WIRES]).collect()

I see you have similar logic in `verify_public_inputs`; I think it would just be slightly nicer if users had one less argument to pass.

Reply:
I'm not sure about this one. I think it can be good to let the user provide the public inputs, and then verify that they are coherent with the proof's openings. Otherwise, it might be too easy to "fool" someone by providing malicious public inputs and hoping no one bothers to check them.